src/wdio/services/SeleniumDockerService.js

```diff
@@ -0,0 +1,125 @@
+import { exec } from 'child_process';
+import retry from 'async/retry';
+import http from 'http';
+import path from 'path';
+import fs from 'fs';
+
+/**
+* Webdriver.io SeleniuMDockerService
+* provides standalone chrome/firefox selenium docker automation.
+*/
+export default class SeleniumDockerService {
+  constructor() {
+    this.getSeleniumStatus = this.getSeleniumStatus.bind(this);
+  }
+
+  getSeleniumStatus(callback) {
+    http.get({
+      host: this.host,
+      port: this.port,
+      path: path.join(this.path || '/wd/hub', 'status'),
+    }, (res) => {
+      const { statusCode } = res;
+      if (statusCode !== 200) {
+        callback('Request failed');
+        return;
+      }
+
+      res.setEncoding('utf8');
+      let rawData = '';
+      res.on('data', (chunk) => { rawData += chunk; });
+      res.on('end', () => {
+        try {
+          const status = JSON.parse(rawData);
+          if (status.value && status.value.ready) {
+            callback(null, status);
+          } else {
+            callback(status);
+          }
+        } catch (e) {
+          callback(`Request failed: ${e.message}`);
+        }
+      });
+    }).on('error', (e) => {
+      callback(`Request failed: ${e.message}`);
+    });
+  }
+
+  /**
+   * Start up docker container before all workers get launched.
+   * @param {Object} config wdio configuration object
+   * @param {Array.<Object>} capabilities list of capabilities details
+   */
+  onPrepare(config, capabilities) {
+    this.config = {
+      cidfile: '.docker_selenium_id', // The docker container id file
+      enabled: true, // True if service enabled, false otherwise
+      cleanup: false, // True if docker container should be removed after test
+      image: null, // The image name to use, defaults to selenium/standalone-${browser}
+      retries: 500, // Retry count to test for selenium being up
+      retryInterval: 10, // Retry interval in milliseconds to wait between retries for selenium to come up.
+      ...(config.seleniumDocker || {}),
+    };
+    this.host = config.host;
+    this.port = config.port;
+    this.path = config.path;
+    this.browserName = capabilities[0].browserName;
+    this.cidfile = this.config.cidfile || this.cidfile;
+
+    return new Promise((resolve, reject) => {
+      if (this.config.enabled) {
+        const containerId = this.getContainerId();
+        if (!containerId) {
+          exec(`docker run --rm --cidfile ${this.cidfile} -p ${config.port}:4444 ${this.getImage()}`);
+        }
+        // Retry for 500 times up to 5 seconds for selenium to start
+        retry({ times: this.config.retries, interval: this.config.retryInterval },
```

Comment:
I ended up getting this running both locally and in jenkins for an internal project.  The one issue that I ran into was that to get it working with our current set up, I needed to know the host that was assigned when the 'docker run' command was executed.  

This wasn't 127.0.0.1 because it was a docker container running (sort of) within another docker container.  I figured out how to determine and update the host dynamically to get this working though.  Not sure if that's something we'll want to do at some point, but I don't think it's worth it for now (unless there are other cases where 127.0.0.1 will not be the host and it can't be predetermined ahead of time).

In the long run I'd like to improve our internal setup so that we can stand up the selenium standalone containers before the test runs even start (similar to how Travis does it and as Mike suggested).  I need to do a little more digging to figure out how to get that working with our set up though.

Here's the jenkins build in case you're curious:  https://jenkins.cerner.com/ion/job/ion-chart/job/ion-chief-complaint-js/job/test_webdriver/46/console

Sorry for the long comment :)

Reply:
no problem, makes sense